File: Data Structures & Algorithms/search-for-word/submission-2.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        
        ROWS, COLS = len(board), len(board[0])
        visited = []
        
        for i in range(ROWS):
            row = []
            for j in range(COLS):
                row.append(False)
            visited.append(row)

        def dfs(r, c, i):

            # Base Case
            if i == len(word):
                return True

            if ( r < 0 or c < 0 or r >= ROWS or c >= COLS
               or word[i] != board[r][c] or visited[r][c] == True):
               return False

            # Recursive Case
            visited[r][c] = True
            res = (dfs(r+1,c,i+1)
                or dfs(r-1,c,i+1)
                or dfs(r,c+1,i+1)
                or dfs(r,c-1,i+1))
            visited[r][c] = False
            return res

        for r in range(ROWS):
            for c in range(COLS):
                if dfs(r,c,0):
                    return True
        return False











            

File: Data Structures & Algorithms/climbing-stairs/submission-2.java
class Solution {
    public int climbStairs(int n) {
        return dfs(n,0);
    }

    private int dfs(int n, int i) {

        // Base case
        if ( i >= n ) return ( i == n ) ? 1 : 0;

        // Recursive case
        return dfs(n, i + 1) + dfs(n, i + 2);

    }
}
